### PR TITLE
SVGLoader: Added fill-opacity style support.

### PR DESCRIPTION
**Pull request by flo-Ty**
Working example here: https://btevc.cf/repos/three.js/examples/webgl_loader_svg_update_1.html

### CODE REVIEW

File: examples/js/loaders/SVGLoader.js
Comment (WestLangley) [inline examples/js/loaders/SVGLoader.js]: formatting: needs spaces -> tabs

**Comment by yomboprime** [comment]
Seems all right to me 👍

File: examples/js/loaders/SVGLoader.js
Comment (BTEVC) [reply examples/js/loaders/SVGLoader.js]: Ah, missed that. Updated now.

**Comment by yomboprime** [comment]
Now that I think on it, the SVGLoader code is adding the `fillOpacity` member to the `ShapePath` instances.
Is that okay?

**Comment by WestLangley** [comment]
I am not sure what is best... Someone who is more-familiar with this will have to speak up.

**Comment by yomboprime** [comment]
(I think @Mugen87 can throw light on this.)

According to mrdoob: https://github.com/mrdoob/three.js/pull/15170#issuecomment-436099412

It seems the `SVGLoader` should not export the SVG tags, but better the SVG node itself, in the object `ShapePath.userData`.

So the SVGLoader should put `path.userData = { node: node }`. And `fillOpacity` should be obtained by the user  with `path.userData.node.getAttribute( 'fill-opacity' )` instead of `path.fillOpacity`.

Additionally I think new functions `svgLoader.getRootNode()` and `svgLoader.getNodePathMap()` (map from nodes to paths) could be called (after the SVG is loaded) if the SVG node tree is needed (see linked PR above)

Exporting the SVG tags allows the user to render things like strokes, which are a nice feature to have. For example, the `tiger.svg` has some shape outlines and not-closed paths that have got tags like `stroke-color` and `stroke-width`. They are not rendered currently.

To not break old user code, a new flag `svgLoader.exportOnlyVisible` or `exportOnlyFillShapes`, which defaults to `true`, could be added to let export (when `false`) the stroke paths which have the `fill-color` set to `none` and thus are currently non-visible and are not exported.

If this is done I can also implement a function that generates, given the path points and the stroke SVG properties, the stroke shape (with miter line joining and no linecaps). I can implement later the tags `strokeLinejoin`, `strokeLinecap` and `strokeMiterLimit` among others.

The stroke creation could also be useful for shapes generated from `Font` text.

**Comment by BTEVC** [comment]
Okay, this is a little above my experience level, but I would be willing to put in the work to figure out how to do it if it would be useful.

**Comment by yomboprime** [comment]
Please, let's wait for some feedback. I am no authority here 😊

If my idea is approved, these are the changes:
 - Keep your changes to the `parseStyle()` function, they are all right.
 - Put `path.userData = { node: node };` just after the line `paths.push( path );`.
 - Revert your other changes to the `SVGLoader`.

With only this, you can access in your application any SVG tag as I said: `path.userData.node.getAttribute( 'fill-opacity' )`

Later other tags can be added to the style and I can do a PR for the stroke creation. 

About the node map, you can also add it if you want:
 - Add a new `this.nodeMap = new Map();` instance variable to the constructor.
 - Add a new `this.rootNode = null;` instance variable to the constructor.
 - Put `nodeMap.set( node, path );` just after the line `paths.push( path );`.
 - Put `this.rootNode = xml.documentElement;` just after the line `parseNode( xml.documentElement, { fill: '#000' } );`

That's all. A getter function for the map and another for the rootNode is optional.

**Comment by WestLangley** [comment]
>To not break old user code

I wouldn't worry about that in this case. You can always add a migration note.

>Please, let's wait for some feedback. I am no authority here

...and I'll defer to @mrdoob and @Mugen87. :-)

**Comment by Mugen87** [comment]
The suggestion of @mrdoob was to introduce `userData` to `ShapePath` which is also done in other classes of `three.js`. Exposing the entire node is probably the most flexible approach rather than exposing individual node properties over `ShapePath`. 

The problems was that the OP of #15170 never implemented his feedback. This PR could be a good opportunity to finally do so 😉

**Comment by BTEVC** [comment]
Okay great, I'll get started on implementing the changes @yomboprime suggested in the next couple of days then.

**Comment by yomboprime** [comment]
Another thought: As the node properties are not inherited between the nodes, I would also add the `style` to `userData`, i.e.:
`path.userData = { node: node, style: style };`

(The styles do inherit properties from upper nodes)

**Comment by yomboprime** [comment]
Looks good. I got already working the stroke creation. In some days I can make a PR.

Screenshot_20190313_012207

I need to implement `stroke-miterlimit` to limit the miter that goes away in some places as you can see. Will implement also round and bevel joins and also linecaps.

**Comment by BTEVC** [comment]
Awesome. 👍

**Comment by BTEVC** [comment]
Woops, had left in the line `this.rootNode = xml.documentElement;` from the nodeMap suggestion. The final decision was to exclude the nodeMap correct?

**Comment by yomboprime** [comment]
Even if the map is out, the root node can be useful. I would leave it. (And declare it in the constructor also, please)

**Comment by BTEVC** [comment]
K, how does that look?

**Comment by BTEVC** [comment]
K, how does that look?

**Comment by yomboprime** [comment]
All right. You keep toggling code 😊

**Comment by BTEVC** [comment]
Haha, yeah, sorry, had a long day today. :)

**Review by Mugen87 — APPROVED** [review]

File: examples/js/loaders/SVGLoader.js
Comment (mrdoob) [inline examples/js/loaders/SVGLoader.js]: Instead of adding this here, I would change `parse()` so it returns an object instead of the current `paths` array.

```js
var data = { paths: [], xml: xml.documentElement };
```

**Comment by mrdoob** [comment]
@WestLangley @Mugen87 @yomboprime thanks for the help here guys!

**Comment by mrdoob** [comment]
@yomboprime these strokes are looking great!

**Comment by yomboprime** [comment]
> @yomboprime these strokes are looking great!

I'm making progress... 😃
strokes_20190315_022443

**Comment by yomboprime** [comment]
I've got the strokes working right.

It is a long function (720 js lines) in `THREE.SVGLoader.pointsToStroke( points, style, arcDivisions )` which returns a BufferGeometry.

I could move it to another place like its own class `StrokeGeometry` in `examples/js/geometries/`. What do you think?

**Comment by mrdoob** [comment]
@yomboprime I was expecting that 😁 I would add it to SVGLoader so it's self contained

**Comment by yomboprime** [comment]
> @yomboprime I was expecting that grin I would add it to SVGLoader so it's self contained

Do you mean that I have a verbose coding habit? 😃

File: examples/js/loaders/SVGLoader.js
Comment (BTEVC) [reply examples/js/loaders/SVGLoader.js]: Okay, updated, does that look right?

**Comment by yomboprime** [comment]
@BTEVC You should also modify the svgloader example for the new change in returned data.

**Comment by BTEVC** [comment]
@yomboprime Oh right, so far only updated my own [example](https://btevc.tk/repos/three.js/examples/webgl_loader_svg_update_1.html).

**Comment by mrdoob** [comment]
> Do you mean that I have a verbose coding habit? 😃

Not at all! I just suspected converting strokes to geometry was going to be arduous 🙈

**Comment by yomboprime** [comment]
@BTEVC You should do:
```
loader.load( url, function ( data ) {
var paths = data.paths;
```
And leave the `for` loop as it was before

**Comment by BTEVC** [comment]
@yamboprime Ah, mkay. Will fix this shortly.

File: examples/webgl_loader_svg.html
Comment (mrdoob) [inline examples/webgl_loader_svg.html]: This seems over complicated. You should be able to keep the previous code and just add a these lines 
```js
loader.load( url, function ( data ) {
```
```js
var paths = data.paths;

for ( var i = 0; i < paths.length; i ++ ) {
```

**Comment by BTEVC** [comment]
So checks still not passing, anyone know what might be going on here?

**Comment by yomboprime** [comment]
> So checks still not passing, anyone know what might be going on here?

It seems something related to webvr renderer files. I have no Idea, could be a Github bug.

**Comment by yomboprime** [comment]
Bonus: Stroke `Font` text. I copied the `webgl_geometry_text_shapes` example and converted the line text to stroke text with minor changes.
strokesText

**Comment by Mugen87** [comment]
> So checks still not passing, anyone know what might be going on here?

Don't worry about that. The travis build was  temporarily broken. Should be fixed now.

**Comment by yomboprime** [comment]
@Mugen87 Can I make my PR if it depends on this one? I'm not a git expert.

**Comment by Mugen87** [comment]
I guess it's better to wait until @mrdoob approves this PR. It has to merged before yours in any event.

BTW: This PR here needs to be mentioned in the migration guide since the parameter of the `onLoad()` function changes. User will have to change their code similar to `webgl_loader_svg`.

**Comment by mrdoob** [comment]
Thanks!

**Comment by BTEVC** [comment]
@yomboprime @WestLangley @Mugen87 @mrdoob Thank you all for helping me through this PR, exciting to participate in this library, even if in a very small way.

**Pull request merged**
